File: Solution Ordered Set.java

import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;

public class StockPrice {

    private final Map<Integer, Integer> timestampToPrice;
    private final TreeMap<Integer, Integer> priceToFrequency;
    private int mostRecentTimestamp;

    public StockPrice() {
        timestampToPrice = new HashMap<>();
        priceToFrequency = new TreeMap<>();
    }

    public void update(int timestamp, int price) {
        if (timestampToPrice.containsKey(timestamp)) {
            int previousPriceForTimestamp = timestampToPrice.get(timestamp);
            int updatedFrequencyForPreviousPrice = priceToFrequency.get(previousPriceForTimestamp) - 1;

            if (updatedFrequencyForPreviousPrice == 0) {
                priceToFrequency.remove(previousPriceForTimestamp);
            } else {
                priceToFrequency.put(previousPriceForTimestamp, updatedFrequencyForPreviousPrice);
            }
        }

        priceToFrequency.put(price, priceToFrequency.getOrDefault(price, 0) + 1);
        timestampToPrice.put(timestamp, price);
        mostRecentTimestamp = Math.max(mostRecentTimestamp, timestamp);
    }

    public int current() {
        return timestampToPrice.get(mostRecentTimestamp);
    }

    public int maximum() {
        return priceToFrequency.lastKey();
    }

    public int minimum() {
        return priceToFrequency.firstKey();
    }
}

File: Solution Ordered Set.cpp

#include <map>
#include <unordered_map>
using namespace std;

class StockPrice {
    
    unordered_map<int, int> timestampToPrice;
    map<int, int, greater<>> priceToFrequency;
    int mostRecentTimestamp{};
    
public:
    StockPrice() = default;

    void update(int timestamp, int price) {

        //C++20: timestampToPrice.contains(timestamp)
        if (timestampToPrice.find(timestamp) != timestampToPrice.end()) {
            int previousPriceForTimestamp = timestampToPrice[timestamp];
            int updatedFrequencyForPreviousPrice = priceToFrequency[previousPriceForTimestamp] - 1;

            if (updatedFrequencyForPreviousPrice == 0) {
                priceToFrequency.erase(previousPriceForTimestamp);
            } else {
                priceToFrequency[previousPriceForTimestamp] = updatedFrequencyForPreviousPrice;
            }
        }

        ++priceToFrequency[price];
        timestampToPrice[timestamp] = price;
        mostRecentTimestamp = max(mostRecentTimestamp, timestamp);
    }

    int current() {
        return timestampToPrice[mostRecentTimestamp];
    }

    int maximum() {
        return priceToFrequency.begin()->first;
    }

    int minimum() {
        return priceToFrequency.rbegin()->first;
    }
};
